File: completable-reactor-runtime/src/main/java/ru/fix/completable/reactor/runtime/internal/dsl/BuilderContext.java
package ru.fix.completable.reactor.runtime.internal.dsl;

import lombok.Data;
import lombok.experimental.Accessors;
import lombok.extern.slf4j.Slf4j;
import ru.fix.completable.reactor.graph.internal.FieldNameResolver;
import ru.fix.completable.reactor.runtime.dsl.MergePoint;
import ru.fix.completable.reactor.runtime.dsl.Processor;
import ru.fix.completable.reactor.runtime.dsl.Subgraph;
import ru.fix.completable.reactor.runtime.internal.CRReactorGraph;
import ru.fix.completable.reactor.runtime.validators.GraphValidator;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * @author Kamil Asfandiyarov
 */
@Slf4j
@Data
@Accessors(chain = true)
public class BuilderContext<PayloadType> {
    final List<GraphValidator> graphValidators = new ArrayList<>();
    final Object graphConfig;
    final CRReactorGraph<PayloadType> graph;
    final FieldNameResolver configFieldNameResolver;

    public BuilderContext(Object graphConfig, CRReactorGraph<PayloadType> graph) {
        this.graphConfig = graphConfig;
        this.graph = graph;
        configFieldNameResolver = new FieldNameResolver(
                graphConfig,
                Arrays.asList(Processor.class, Subgraph.class, MergePoint.class));
    }

    public String resolveConfigField(Object configField) {
        return configFieldNameResolver.resolveFieldName(configField);
    }
}

File: completable-reactor-graph/src/main/kotlin/ru/fix/completable/reactor/graph/internal/BuilderContext.kt
package ru.fix.completable.reactor.graph.internal

import ru.fix.completable.reactor.graph.Graphable
import ru.fix.completable.reactor.graph.Vertex
import ru.fix.completable.reactor.graph.runtime.RuntimeGraph

class BuilderContext {

    private var fieldNameResolver: FieldNameResolver? = null

    private var graph: RuntimeGraph? = null

    fun setGraph(graph: Graphable, runtimeGraph: RuntimeGraph) {
        this.graph = runtimeGraph
        this.fieldNameResolver = FieldNameResolver(graph, listOf(Vertex::class.java))
    }

    fun getGraph() = graph

    companion object {
        private val builderContext = ThreadLocal<BuilderContext>()

        @JvmStatic
        fun get(): BuilderContext {
            var context: BuilderContext? = builderContext.get()
            if (context == null) {
                context = BuilderContext()
                builderContext.set(context)
            }
            return context
        }
    }

    fun resolveVertexName(vertex: Vertex): String? {
        return fieldNameResolver?.resolveFieldName(vertex)
    }
}
